Objects/frameobject.c

```diff
@@ -879,20 +879,21 @@ frame_dealloc(PyFrameObject *f)
     /* It is the responsibility of the owning generator/coroutine
      * to have cleared the generator pointer */
 
-    assert(f->f_frame->owner != FRAME_OWNED_BY_GENERATOR ||
-        _PyFrame_GetGenerator(f->f_frame)->gi_frame_state == FRAME_CLEARED);
-
     if (_PyObject_GC_IS_TRACKED(f)) {
         _PyObject_GC_UNTRACK(f);
     }
 
     Py_TRASHCAN_BEGIN(f, frame_dealloc);
     PyObject *co = NULL;
 
+    /* GH-106092: If f->f_frame was on the stack and we reached the maximum
+     * nesting depth for deallocations, the trashcan may have delayed this
+     * deallocation until after f->f_frame is freed. Avoid dereferencing
+     * f->f_frame unless we know it still points to valid memory. */
+    _PyInterpreterFrame *frame = (_PyInterpreterFrame *)f->_f_frame_data;
+
     /* Kill all local variables including specials, if we own them */
-    if (f->f_frame->owner == FRAME_OWNED_BY_FRAME_OBJECT) {
-        assert(f->f_frame == (_PyInterpreterFrame *)f->_f_frame_data);
-        _PyInterpreterFrame *frame = (_PyInterpreterFrame *)f->_f_frame_data;
+    if (f->f_frame == frame && frame->owner == FRAME_OWNED_BY_FRAME_OBJECT) {
```

Comment:
This looks correct.

Note:
In theory, `f->f_frame == frame` and `frame->owner == FRAME_OWNED_BY_FRAME_OBJECT` should be equivalent, but in practice they are not.
We can't just check `frame->owner == FRAME_OWNED_BY_FRAME_OBJECT` because of the Trashcan mechanism, and we can't just check `f->f_frame == frame` because of [this](https://github.com/python/cpython/blob/e8a16a6dae7e20c65a96cc3dec1dc38914caba02/Python/frame.c#L40).
So we need both checks here.

Reply:
Should this check be put in other places too? We are currently investigating the root cause of segfaults in 3.11.5 with call stacks ending with
```
#0  0x0000000000635d9a in ?? ()
#1  0x00000000004d8e27 in PyFrame_FastToLocalsWithError ()
#2  0x00000000004d8e3e in ?? ()
#3  0x000000000054e5f6 in _PyObject_GenericGetAttrWithDict ()
```
What we are trying to access here is `frame.f_locals`. The frame object is obtained from looping over all thread states and calling `PyThreadState_GetFrame` on them (so we know we have a strong reference to it). Because of the inlining it's not immediately obvious where in `PyFrame_FastToLocalsWithError` we are segfaulting exactly, but our current hunch is that it is because the referenced `_PyInterpreterState` might have been deallocated. We're still investigating this.